{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib notebook\n",
    "%precision 3\n",
    "\n",
    "from IPython.display import set_matplotlib_formats\n",
    "from pandas.plotting import register_matplotlib_converters\n",
    "register_matplotlib_converters()\n",
    "\n",
    "import matplotlib\n",
    "import math\n",
    "import numpy as np\n",
    "import seaborn as sb\n",
    "import matplotlib.pyplot as plt\n",
    "plt.rcParams['figure.figsize'] = [9.5, 7]\n",
    "import pandas\n",
    "from random import shuffle\n",
    "from scipy.stats import poisson\n",
    "import statsmodels.api as sm\n",
    "import statsmodels.formula.api as smf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set this to where you placed https://berthub.eu/antonie/gcskewdb.csv.gz and unpacked https://berthub.eu/antonie/gcfits.tar.bz2\n",
    "prefix=\"/home/ahu/git/antonie/genomes/ncbi-genomes-2021-08-19/\"\n",
    "m=pandas.read_csv(prefix+\"gcskewdb.csv\")\n",
    "m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chromoname=\"NZ_CP019870.1\" # c difficile\n",
    "fitted = pandas.read_csv(prefix+\"/\"+chromoname+\"_fit.csv\")\n",
    "\n",
    "us=m[m.name==chromoname]\n",
    "\n",
    "plt.figure()\n",
    "plt.plot(fitted.pos, fitted.gcskew, label=\"Cumulative GC skew\")\n",
    "plt.plot(fitted.pos, fitted.predgcskew, label=\"Fitted GC skew\")\n",
    "#plt.plot(fitted.pos, fitted.taskew, label=\"Cumulative TA skew\")\n",
    "#plt.plot(fitted.pos, fitted.predtaskew, label=\"Fitted TA skew\")\n",
    "\n",
    "#plt.plot(fitted.pos, fitted.sbskew, label=\"Cumulative SB skew\")\n",
    "#plt.plot(fitted.pos, fitted.predsbskew, label=\"Fitted SB skew\")\n",
    "\n",
    "\n",
    "leshift=us[\"shift\"].item()\n",
    "if leshift > 0:\n",
    "    plt.axvline(leshift, ls=':', color='black')\n",
    "else:\n",
    "    plt.axvline(fitted.pos.max() + leshift, ls=':', color='black')\n",
    "\n",
    "plt.xlabel(\"Locus\")\n",
    "plt.ylabel(\"Skew\")\n",
    "plt.legend()\n",
    "plt.title(chromoname + \" alpha1 \" + str(round(us[\"alpha1gc\"].item(),3)) + \" alpha2 \" + str(round(us[\"alpha2gc\"].item(),3)) + \n",
    "          \" div \" + str(round(us[\"div\"].item(),3)))\n",
    "plt.grid()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure()\n",
    "\n",
    "leX=fitted.sbskew.diff().rolling(10, center=True).mean()\n",
    "leY=fitted.gcskew.diff().rolling(10, center=True).mean()\n",
    "\n",
    "\n",
    "frame = { 'predleading': fitted.predleading, 'x': leX, 'y': leY }\n",
    "result = pandas.DataFrame(frame)\n",
    "#print(result.tail(20))\n",
    "\n",
    "\n",
    "sub=result[(result.predleading==1) & ((result.x < 0) | (result.x>0)) & ((result.y < 0) | (result.y > 0)) ]\n",
    "#z = np.polyfit(sub.x, sub.y, 1)\n",
    "#p = np.poly1d(z)\n",
    "#xp = np.linspace(leX.min(), leX.max(), 100)\n",
    "#plt.plot(xp, p(xp), ':', color='red')\n",
    "\n",
    "model = smf.quantreg('y ~ x', sub).fit(q=0.5)\n",
    "print(model.summary())\n",
    "\n",
    "print(model.params['Intercept'], model.params['x'])\n",
    "\n",
    "get_y = lambda a, b: a + b * sub.x\n",
    "y = get_y(model.params['Intercept'], model.params['x'])\n",
    "plt.plot(sub.x, y, color='black')\n",
    "\n",
    "\n",
    "sub=result[(result.predleading==0) & ((result.x < 0) | (result.x>0)) & ((result.y < 0) | (result.y > 0))]\n",
    "\n",
    "#z = np.polyfit(sub.x, sub.y, 1)\n",
    "#p = np.poly1d(z)\n",
    "#xp = np.linspace(leX.min(), leX.max(), 100)\n",
    "#plt.plot(xp, p(xp), ':', color='red')\n",
    "\n",
    "model = smf.quantreg('y ~ x', sub).fit(q=0.5)\n",
    "print(model.summary())\n",
    "print(model.params['Intercept'], model.params['x'])\n",
    "get_y = lambda a, b: a + b * sub.x\n",
    "y = get_y(model.params['Intercept'], model.params['x'])\n",
    "plt.plot(sub.x, y, color='black')\n",
    "\n",
    "\n",
    "plt.scatter(result[result.predleading==0].x, result[result.predleading==0].y, marker=\"+\", color='blue', label=\"GC lagging\")\n",
    "plt.scatter(result[result.predleading==1].x, result[result.predleading==1].y, marker=\"^\", color='blue', label=\"GC leading\")\n",
    "\n",
    "leX=fitted.sbskew.diff().rolling(10, center=True).mean()\n",
    "leY=fitted.taskew.diff().rolling(10, center=True).mean()\n",
    "\n",
    "z = np.polyfit(leX.dropna(), leY.dropna(), 1)\n",
    "p = np.poly1d(z)\n",
    "xp = np.linspace(leX.min(), leX.max(), 100)\n",
    "plt.plot(xp, p(xp), ':', color='red')\n",
    "\n",
    "plt.scatter(leX, leY,  s=4, label=\"TA\", color='orange')\n",
    "plt.xlabel(\"SB skew\")\n",
    "plt.ylabel(\"GC/TA skew\")\n",
    "plt.title(chromoname)\n",
    "plt.legend()\n",
    "\n",
    "\n",
    "plt.grid()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# find and plot a bunch of flat chromosomes\n",
    "\n",
    "sel=m[(m.rmsGC < 0.0020) & (m[\"alpha1gc\"] < 0.0018) & (m[\"alpha1gc\"] > 0)]  #  & (m.realm3==\"Firmicutes\")\n",
    "#sel=m[(m[\"div\"] < 0.22) & (m.rmsGC < 0.002) ]\n",
    "d=4\n",
    "fig,axs=plt.subplots(d,d, sharex=False, sharey=False, constrained_layout=True)\n",
    "a=0\n",
    "b=0\n",
    "names=sel.name.unique()\n",
    "print(len(names))\n",
    "shuffle(names)\n",
    "for k in names:\n",
    "        fitted = pandas.read_csv(prefix+\"/\"+k+\"_fit.csv\")\n",
    "        t=m[m.name==k]\n",
    "        axs[b,a].plot(fitted.pos, fitted.gcskew)\n",
    "        axs[b,a].plot(fitted.pos, fitted.predgcskew)\n",
    "        #axs[b,a].plot(fitted.pos, fitted.taskew)\n",
    "        #axs[b,a].plot(fitted.pos, fitted.predtaskew)\n",
    "\n",
    "        #axs[b,a].get_yaxis().set_ticks([])\n",
    "        axs[b,a].set_title(k, fontsize=8) #  + \" \"+str(1000*results[results.name==k].rmsGC.mean()), \n",
    "        #axs[b,a].grid()\n",
    "        a=a+1\n",
    "        if(a>=d):\n",
    "            a=0\n",
    "            b=b+1\n",
    "        if(b == d):\n",
    "            break\n",
    "fig.suptitle(\"GC/TA skew in random bacterial chromosomes + fit\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot relation of GC and TA skew for Firmicutes and non-Firmicutes\n",
    "plt.figure()\n",
    "sel=m[m.realm3==\"Firmicutes\"]\n",
    "plt.scatter(sel.alpha1gc, sel.alpha1ta, s=2, alpha=0.3, label=\"Firmicutes\")\n",
    "sel=m[m.realm3!=\"Firmicutes\"]\n",
    "plt.scatter(sel.alpha1gc, sel.alpha1ta, s=2, alpha=0.3, label=\"Non-Firmicutes\")\n",
    "\n",
    "plt.xlabel(\"alpha1 of GC skew\")\n",
    "plt.ylabel(\"alpha1 of TA skew\")\n",
    "plt.legend()\n",
    "plt.title(\"Scatter plot of GC and TA skew\")\n",
    "plt.grid()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot prediction of GC/TA skew based on strand and codon bias and percentage non-coding\n",
    "firmi=m[(m.realm3==\"Firmicutes\") & (m.rmsGC < 0.02) & (m.rmsTA < 0.02)]\n",
    "plt.figure()\n",
    "leX =  -(firmi.cfrac - firmi.gfrac)*(1-firmi.ngcount/firmi.siz)*firmi.alpha2sb\n",
    "\n",
    "leY=firmi.alpha1gc\n",
    "\n",
    "frame = {  'x': leX, 'y': leY }\n",
    "sub = pandas.DataFrame(frame)\n",
    "\n",
    "model = smf.quantreg('y ~ x', sub).fit(q=0.5)\n",
    "print(model.summary())\n",
    "\n",
    "print(model.params['Intercept'], model.params['x'])\n",
    "\n",
    "get_y = lambda a, b: a + b * sub.x\n",
    "y = get_y(model.params['Intercept'], model.params['x'])\n",
    "\n",
    "plt.scatter(leX, leY, s=2, label=\"GC\")\n",
    "plt.plot(sub.x, y, color='black')\n",
    "\n",
    "#\n",
    "\n",
    "leX =  -(firmi.tfrac - firmi.afrac)*(1-firmi.ngcount/firmi.siz)*firmi.alpha2sb\n",
    "leY=firmi.alpha1ta\n",
    "\n",
    "frame = {  'x': leX, 'y': leY }\n",
    "sub = pandas.DataFrame(frame)\n",
    "\n",
    "model = smf.quantreg('y ~ x', sub).fit(q=0.5)\n",
    "print(model.summary())\n",
    "\n",
    "print(model.params['Intercept'], model.params['x'])\n",
    "\n",
    "get_y = lambda a, b: a + b * sub.x\n",
    "y = get_y(model.params['Intercept'], model.params['x'])\n",
    "plt.scatter(leX, leY, s=2, label=\"TA\")\n",
    "\n",
    "plt.plot(sub.x, y, color='black', label='fit')\n",
    "\n",
    "\n",
    "plt.grid()\n",
    "#plt.scatter(firmi.ccounts2-firmi.tcounts2, firmi.alpha1gc)\n",
    "plt.legend()\n",
    "plt.xlabel(\"Product of gene strand bias, codon bias skew, percentage coding\")\n",
    "plt.ylabel(\"GC/TA skew fraction\")\n",
    "plt.title(\"Data for \"+str(len(firmi))+ \" Firmicute chromosomes\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
